### ISSUE TITLE
How to see the Fourier modes being learned

### ISSUE THREAD

**Issue by suyashbire1** [neuraloperator/neuraloperator#146]
Hi,

Thank you for this library!

I am wondering if there is an easy way to extract the Fourier modes being learnt. I am hoping to extract the state of the tensor, x, immediately after the inverse fft operation for each layer.

For example, here's a simple 2D FNO model

```python
model = FNO(n_modes=(64, 64), n_layers=3, hidden_channels=64, in_channels=11, out_channels=10)
       
print(model)

FNO(
  (fno_blocks): FNOBlocks(
    (convs): FactorizedSpectralConv(
      (weight): ModuleList(
        (0): ComplexDenseTensor(shape=torch.Size([64, 64, 32, 32]), rank=None)
        (1): ComplexDenseTensor(shape=torch.Size([64, 64, 32, 32]), rank=None)
        (2): ComplexDenseTensor(shape=torch.Size([64, 64, 32, 32]), rank=None)
        (3): ComplexDenseTensor(shape=torch.Size([64, 64, 32, 32]), rank=None)
        (4): ComplexDenseTensor(shape=torch.Size([64, 64, 32, 32]), rank=None)
        (5): ComplexDenseTensor(shape=torch.Size([64, 64, 32, 32]), rank=None)
      )
    )
    (fno_skips): ModuleList(
      (0): Conv2d(64, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (1): Conv2d(64, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (2): Conv2d(64, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
    )
  )
  (lifting): Lifting(
    (fc): Conv2d(11, 64, kernel_size=(1, 1), stride=(1, 1))
  )
  (projection): Projection(
    (fc1): Conv2d(64, 256, kernel_size=(1, 1), stride=(1, 1))
    (fc2): Conv2d(256, 10, kernel_size=(1, 1), stride=(1, 1))
  )
)
```

I can see the output from the skip modules but not from the conv modules:
```python
features = {}
def get_features(name):
    def hook(model, input, output):
        features[name] = output.detach()
    return hook

for block in range(3):
    model.fno_blocks.convs[block].register_forward_hook(get_features(f'postfnob{block}'))          # doesn't work
    model.fno_blocks.fno_skips[block].register_forward_hook(get_features(f'postfnosb{block}'))     # works!
```
Is there an easy way to see the output of convs?

**Comment by itskobold** [comment]
As far as I'm aware this isn't supported in the library yet but I'm working on adding it myself - will share once it's complete ofc.

**Comment by JeanKossaifi** [comment]
That one is slightly subtle given that we allow for joint parameterizations.
You should be able to do 
```
model.fno_blocks.convs.weight[block]
```
or 
```
model.fno_blocks.convs
```
knowing that a Spectral conv parametrizes more than one layer.

**Comment by JeanKossaifi** [comment]
Is this what you need @suyashbire1 ?

**Comment by suyashbire1** [comment]
If I understand it correctly, `_contract_dense` extracts the complex weight tensor for a particular block which is then multiplied to the Fourier decomposition to get a new Fourier modal space. So we should be able to recreate this new Fourier space if we know the input tensor `x` and the weight given by `_contract_dense`. Is this correct @JeanKossaifi ?

**Comment by JeanKossaifi** [comment]
Exactly @suyashbire1 

The code supports multiple ways of parametrizing the spectral convolution - namely, you can have the weights be "dense" or in factorized form. Regardless, if you use `implementation='reconstructed'`, you can always i) reconstruct the dense weights from the factorized ones and ii) contract those with your inputs `x` as you are saying so indeed, in that case you can directly look at these in `_contract_dense`.

 Hope this answers your question!

**Comment by suyashbire1** [comment]
For my particular use case I have `norm=True`. Normalization layers allow registering forward hooks. So it is possible to get outputs after normalization layers.

**Comment by JeanKossaifi** [comment]
Great! Closing, feel free to reopen if you have more questions.
